### PR TITLE
Branch setup

### PR DESCRIPTION
**Pull request by milismsft**
Merge latest from master
Add Application Gateway ops (create, delete)
Add Application Gateway caching
Add Application gateway viewing via new HTTP request mapping (see localhost:7002/azure/appgateways). Expect this to change once we have all the pieces in place to switch over to using application gateways instead of load balancers.

### CODE REVIEW

**Comment by milismsft** [comment]
@scotmoor @JargoonPard PTAL

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy [hunk 1]
```diff
@@ -243,6 +247,124 @@ class AzureNetworkClient extends AzureBaseClient {
...
+        agItem.tags = appGateway.tags
+        agItem.dnsName = getDnsNameForAppGateway(AzureUtilities.getResourceGroupNameFromResourceId(appGateway.id), appGateway.name)
+        agItem.lastReadTime = currentTime
+        return agItem
+      }
+    } catch (CloudException e) {
+      log.error("getAppGateway(${resourceGroupName},${appGatewayName}) -> Cloud Exception ", e)
+    } catch (Exception e) {
+      log.error("getAppGateway(${resourceGroupName},${appGatewayName}) -> Exception ", e)
+    }
```
Comment (scotmoor) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy:258]: There is no need for two different 'catch' clauses. The output will show what kind of exception it actually is.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy [hunk 2]
```diff
@@ -243,6 +247,124 @@ class AzureNetworkClient extends AzureBaseClient {
...
+   * get the dns name associated with an Application Gateway in Azure
+   * @param resourceGroupName name of the resource group where the application gateway was created (see application name and region/location)
+   * @param appGatewayName the name of the application gateway in Azure
+   * @return the dns name of the Public IP for the given application gateway or "dns-not-found"
+   */
+  String getDnsNameForAppGateway(String resourceGroupName, String appGatewayName) {
+    String dnsName = "dns-not-found"
+
+    try {
+      def appGateway = executeOp({appGatewayOps.get(resourceGroupName, appGatewayName)})?.body
```
Comment (scotmoor) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy]: In both instances where you call this method, you've just retrieved the App Gateway from Azure via the list operation. Why do you need to retrieve again?

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy [hunk 3]
```diff
@@ -572,7 +694,7 @@ class AzureNetworkClient extends AzureBaseClient {
    * @return the dns name of the given load balancer
    */
   String getDnsNameForLoadBalancer(String resourceGroupName, String loadBalancerName) {
-    String dnsName = "none"
+    String dnsName = "dns-not-found"
 
     try {
       def loadBalancer = executeOp({loadBalancerOps.get(resourceGroupName, loadBalancerName, null)})?.body
```
Comment (scotmoor) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy]: Same as getting dns for app gateway. We've already read the item out of Azure. We shouldn't be reading it out again.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy (hunk 2 above)
Comment (milismsft) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy]: It makes it more generic and can be called just with the name of the application gateway (rather than passing an Azure ApplicationGateway resource).

Comment (scotmoor) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy]: Can you please explain how a method whose sole purpose is to get the DNS name for an App Gateway is generic? Also, I'm not sure how passing the app gateway name instead of the resource you've read from Azure makes any difference.  Finally, and this is a big one, you've unnecessarily introduced a perf issue.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/cache/AzureAppGatewayCachingAgent.groovy [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+import com.netflix.spinnaker.clouddriver.azure.security.AzureCredentials
+import com.netflix.spinnaker.clouddriver.cache.OnDemandAgent
+import com.netflix.spinnaker.clouddriver.cache.OnDemandMetricsSupport
+import groovy.transform.WithWriteLock
+import groovy.util.logging.Slf4j
+
+import static com.netflix.spinnaker.cats.agent.AgentDataType.Authority.AUTHORITATIVE
+
+@Slf4j
+class AzureAppGatewayCachingAgent implements CachingAgent, OnDemandAgent, AccountAware{
```
Comment (scotmoor) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/cache/AzureAppGatewayCachingAgent.groovy:45]: Is there a particular reason why you chose not to derive this from the AzureCachingAgent? You're duplicating code unecessarily

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/view/AzureAppGatewayProvider.groovy [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    AzureAppGatewayDescription description = objectMapper.convertValue(cacheData.attributes['appgateway'], AzureAppGatewayDescription)
+    def parts = Keys.parse(azureCloudProvider, cacheData.id)
+
+    def loadBalancer = new AzureLoadBalancer(
+      account: parts.account ?: "none",
+      name: description.loadBalancerName,
+      region: description.region,
+      vnet: description.vnet ?: "vnet-unassigned",
+      subnet: description.subnet ?: "subnet-unassigned",
+      cluster: description.cluster ?: "unassigned"
```
Comment (scotmoor) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/view/AzureAppGatewayProvider.groovy:82]: There's no cluster property in an AzureLoadBalancer

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/view/AzureAppGatewayController.groovy [hunk 6]
```diff
@@ -0,0 +1,140 @@
+/*
```
Comment (scotmoor) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/view/AzureAppGatewayController.groovy:1]: Why is this class in this PR?

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/view/AzureAppGatewayProvider.groovy [hunk 7]
```diff
@@ -0,0 +1,110 @@
+/*
```
Comment (scotmoor) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/view/AzureAppGatewayProvider.groovy:1]: Same. Why is this included?

File: clouddriver-azure/src/test/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/deploy/ops/DeleteAzureAppGatewayAtomicOperationSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+  def setupSpec() {
+    this.converter = new DeleteAzureAppGatewayAtomicOperationConverter(objectMapper: mapper)
+    def accountCredentialsProvider = Mock(AccountCredentialsProvider)
+    def mockCredentials = Mock(AzureNamedAccountCredentials)
+    accountCredentialsProvider.getCredentials(_) >> mockCredentials
+    converter.accountCredentialsProvider = accountCredentialsProvider
+  }
+
+  void "Create UpsertAzureAppGatewayAtomicOperation object - simple test"() {
```
Comment (scotmoor) [inline clouddriver-azure/src/test/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/deploy/ops/DeleteAzureAppGatewayAtomicOperationSpec.groovy]: Should this read "DeleteAzureAppGatewayAtomicOperation object"?

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/view/AzureAppGatewayProvider.groovy (hunk 7 above)
Comment (milismsft) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/view/AzureAppGatewayProvider.groovy:1]: these classes are needed in order to test the implementation from clouddriver HTTP entry points only (see my PR comment). Basically you can see the AppGateways and do some of the AppGateway related ops via localhost:7002 and curl commands before the switch from using lb to ag.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/view/AzureAppGatewayProvider.groovy (hunk 5 above)
Comment (milismsft) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/view/AzureAppGatewayProvider.groovy:82]: sorry, lost that during the merge last night. I will add it back.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/cache/AzureAppGatewayCachingAgent.groovy (hunk 4 above)
Comment (milismsft) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/cache/AzureAppGatewayCachingAgent.groovy:45]: None in particular; the implementation follows the existing Load Balancer one since I know it works 100% of the time. That said once this work is done there's some consolidation work that should follow, not just of AppGateway but also security groups. Feel free to add a Trello card for that.

File: clouddriver-azure/src/test/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/deploy/ops/DeleteAzureAppGatewayAtomicOperationSpec.groovy (hunk 8 above)
Comment (milismsft) [inline clouddriver-azure/src/test/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/deploy/ops/DeleteAzureAppGatewayAtomicOperationSpec.groovy]: yep

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy (hunk 2 above)
Comment (milismsft) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy]: If you noticed, except couple methods in the AzureNetworkClient that deal with constructing the description objects, all other take the name of the resources to operate on. That said, in this particular case I think I can move the code into the get/getAll methods (one less method to deal with).

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/cache/AzureAppGatewayCachingAgent.groovy (hunk 4 above)
Comment (scotmoor) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/cache/AzureAppGatewayCachingAgent.groovy:45]: Make the change to derive this class from AzureCachingAgent now.  Do not make unnecessary work that we may or may not get to.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy [hunk 9]
```diff
@@ -210,7 +210,7 @@ public class AzureComputeClient extends AzureBaseClient {
   Collection<AzureInstance> getServerGroupInstances(String resourceGroupName, String serverGroupName) {
     def instances = new ArrayList<AzureInstance>()
 
-    executeOp({scaleSetVMOps.list(resourceGroupName, serverGroupName, null, null, "instanceView")}).body.each {
+    executeOp({scaleSetVMOps.list(resourceGroupName, serverGroupName, null, null, "instanceView")})?.body?.each {
```
Comment (milismsft) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy:213]: This is one of the cases where we blew up trying to retrieve the resources associated with the server group in the server group caching agent.

**Comment by milismsft** [comment]
@scotmoor @JargoonPard PTAL
